source/agora/consensus/Validation.d

```diff
@@ -990,6 +993,70 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
     return null;
 }
 
+/*******************************************************************************
+
+    Check the validator of an enrollment.
+
+    A Validator's enrollment is considered valid if:
+        - UTXO is unspent frozen utxo
+        - Signatures are authentic
+        - The frozen amount must be equal to or greater than 40,000 BOA
+        - The block height shall be greater than the unlock_height of the UTXO.
+
+    Params:
+        block_height = the height of the direct ancestor of this block
+        enrollment = The enrollment of the target to be verified
+        findUTXO = delegate to find the referenced unspent UTXOs with
+
+    Returns:
+        `null` if the validator's UTXO is valid, a string explaining the reason it
+        is invalid otherwise.
+
+*******************************************************************************/
+
+public string isInvalidEnrollmentReason (const ulong block_height,
+    const ref Enrollment enrollment, UTXOFinder findUTXO) nothrow @safe
+{
+    UTXOSetValue utxo_set_value;
+    if (!findUTXO(enrollment.utxo_key, size_t.max, utxo_set_value))
+        return "Unspent frozen UTXO not found for the validator.";
+
+    if (utxo_set_value.type != TxType.Freeze)
+        return "UTXO is not frozen.";
+
+    bool isVerify = false;
+    try
+    {
+        const Point address = Point(utxo_set_value.output.address);
```

Comment:
We should make the crypto routines @safe or @trusted. This can be done in another PR. 

Also this can be simplified to:

```D
 if (!() @trusted { return verify(address, enrollment.enroll_sig, enrollment); }())
    return "Enrollment signature verification has an error.";
```

This gets rid of a state variable. And also `isVerify` is not according to our style guide, it would have to be `is_verify` or better yet `is_valid`.

Also, if possible this call should be moved outside the `try/catch` because `verify` is `nothrow`.

Reply:
applied the suggest.

Reply:
If you rebase your PR, you can remove the `@trusted` workaround, because https://github.com/bpfkorea/agora/pull/466 was merged.

Reply:
Applied PR #466.